finality-aleph/src/justification/requester.rs

```diff
@@ -32,76 +35,104 @@ impl<B: BlockT> JustificationRequestStatus<B> {
         Self {
             block_number: None,
             block_hash: None,
-            tries: 0,
+            block_tries: 0,
+            parent: None,
+            n_childred: None,
+            children_tries: 0,
             report_threshold: REPORT_THRESHOLD,
         }
     }
 
-    fn save_block_number(&mut self, num: NumberFor<B>) {
-        if Some(num) == self.block_number {
-            self.tries += 1;
+    fn save_children(&mut self, hash: B::Hash, n_childred: usize) {
+        if self.parent == Some(hash) {
+            self.children_tries += 1;
         } else {
-            self.block_number = Some(num);
-            self.block_hash = None;
-            self.tries = 1;
+            self.parent = Some(hash);
+            self.n_childred = Some(n_childred);
+            self.children_tries = 1;
         }
     }
 
-    fn insert_hash(&mut self, hash: B::Hash) {
-        self.block_hash = Some(hash);
+    fn save_block(&mut self, num: NumberFor<B>, hash: B::Hash) {
+        if self.block_number == Some(num) {
+            self.block_tries += 1;
+        } else {
+            self.block_hash = Some(hash);
+            self.block_number = Some(num);
+            self.block_hash = None;
+            self.block_tries = 1;
+        }
     }
 
     fn should_report(&self) -> bool {
-        self.tries >= self.report_threshold
+        self.block_tries >= self.report_threshold || self.children_tries >= self.report_threshold
     }
 }
 
 impl<B: BlockT> fmt::Display for JustificationRequestStatus<B> {
     fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
-        if let Some(n) = self.block_number {
-            let mut status = format!("tries - {}; requested block number - {}; ", self.tries, n);
-            if let Some(header) = self.block_hash {
-                status.push_str(&format!("hash - {}; ", header));
-            } else {
-                status.push_str("hash - unknown; ");
+        if self.block_tries >= self.report_threshold {
+            if let Some(n) = self.block_number {
+                let mut status = format!(
+                    "tries - {}; requested block number - {}; ",
+                    self.block_tries, n
+                );
+                if let Some(header) = self.block_hash {
+                    status.push_str(&format!("hash - {}; ", header));
+                } else {
+                    status.push_str("hash - unknown; ");
+                }
+
+                write!(f, "{}", status)?;
             }
+        }
+        if self.children_tries >= self.report_threshold {
+            if let Some(parent) = self.parent {
+                let n = self
+                    .n_childred
```

Comment:
This could just be usize and by default set to `0` if it is saved with parent. Or if you want to keep option, even:
```
if let (Some(parent), Some(n)) = (self.parent, self.n_children) {
    ...
}
```

Reply:
Removed Option.